### PR TITLE
fix changelog record for circuit termination

### LINKED ISSUE #23134: Circuit termination pointer is updated without a changelog record
### NetBox Edition

NetBox Community

### NetBox Version

v4.7.0

### Python Version

3.12

### Steps to Reproduce

1. Create a circuit.
2. Add an A-side termination to it.
3. Open the circuit's **Changelog** tab.

### Expected Behavior

A change record for the **circuit**, showing `termination_a` going from `null` to the new termination's ID — the field did change, and it is included in the serialized changelog data.

Note: You do get a CircuitTermination changelog (it being created) but no Updated changelog for the actual Circuit.

### Observed Behavior

No change record for the circuit at all. The only record is `created` for the circuit termination. `Circuit.termination_a` is set in the database, but nothing in the changelog reflects it.

Anything that consumes the changelog therefore misses the association: event rules on circuit updates don't fire, and the netbox-branching plugin drops `termination_a`/`termination_z` when a termination created in a branch is merged (netboxlabs/netbox-branching#651).

### Suspected Cause

`CircuitTermination.save()` maintains the pointer with a raw queryset update (`circuits/models/circuits.py`, both the clear-old and set-new branches):

  ```python
  Circuit.objects.filter(pk=self.circuit_id).update(**{termination_name: self.pk})
```

QuerySet.update() emits no post_save, so core.signals.handle_changed_object never runs and no ObjectChange is written.

### Proposed Fix

Route the write through a change-logged path — snapshot, assign, save the single field:
```
circuit = Circuit.objects.filter(pk=circuit_id).first()
if circuit is None or getattr(circuit, f'{field_name}_id') == value:
    return
circuit.snapshot()
setattr(circuit, f'{field_name}_id', value)
circuit.save(update_fields=[field_name, 'last_updated'])
```

Two notes for whoever picks this up: the circuit must be re-fetched rather than reusing a cached self.circuit, or creating the A then Z termination snapshots a stale pre-A-write state; and termination_a/termination_z are editable=False but are still serialized, so the emitted change does carry the value.

(Deleting a termination clears the pointer via on_delete=SET_NULL, which is also a signal-free bulk update and equally unlogged — same field, worth fixing together.)

### PR DESCRIPTION
**Pull request by arthanson**
### Closes: #23134

`CircuitTermination.save()` maintained the parent circuit's `termination_a`/`termination_z` pointers with a queryset `update()`, which emits no `post_save` and so was never change-logged. Deletion was equally invisible: `on_delete=SET_NULL` clears the column with a bulk UPDATE. Consumers replaying `ObjectChange` records drop the association on merge (netboxlabs/netbox-branching#651).

Both paths now fetch the circuit and write the pointer via `snapshot()` + `save(update_fields=[...])`. The fetch matches on the pointer's current value, so it is skipped unless the pointer actually references this termination. `delete()` clears the pointer before the row goes away, so its record precedes the DELETE. Cascades reach the row through the collector and remain unrecorded.

Also added a fix for cascade deletion which similarly didn't update the circuit terminations fields.

**Cost:** each create, move or delete adds a SELECT, a real `Circuit.save()` and an `ObjectChange` row in place of one unlogged UPDATE. `Circuit.last_updated` now advances and `OBJECT_UPDATED` fires, so event rules on circuit updates trigger on termination writes and the circuit's REST `If-Match` ETag is invalidated.

### CODE REVIEW

**Comment by arthanson** [comment]
@claude /review Only report on issues of potential concern; we don't need commentary on anything that doesn't require further consideration.

**Comment by arthanson** [comment]
@claude /review Only report on issues of potential concern; we don't need commentary on anything that doesn't require further consideration.

**Comment by arthanson** [comment]
@claude /review Only report on issues of potential concern; we don't need commentary on anything that doesn't require further consideration.

**Comment by arthanson** [comment]
@claude /review Only report on issues of potential concern; we don't need commentary on anything that doesn't require further consideration.

**Comment by arthanson** [comment]
Pre Claude review:

Points 1 and 4 are PR-description items.

Points 2b, 3, and 5 are fixed;

**Comment by arthanson** [comment]
@claude /review Only report on issues of potential concern; we don't need commentary on anything that doesn't require further consideration.

**Comment by arthanson** [comment]
@claude /review Only report on issues of potential concern; we don't need commentary on anything that doesn't require further consideration.

**Comment by arthanson** [comment]
Per Claude review:

1. Fixed - clear gated on not is_new; test added.
2. Fixed - delete() locks the termination row before the circuit, matching save()'s order.
3. Not fixed - needs to happen if we are doing a save.
4. Fixed - dropped update_fields for a full save, so postchange_data matches the row as written.
5. Fixed - using now read from args[0] as well as kwargs.
6. Not fixed - only cascade+revert remains; bulk_* and queryset.delete() don't appear in replay.

**Review by jeremystretch — CHANGES_REQUESTED** [review]
This seems like quite a bit of bespoke logic for a fairly simple change. Is there anything preventing us from changing the `update()` calls to `save(update_fields=[...])` on the parent Circuit not suffice? Claude also found a number of concerns (below).

---

### 1. `netbox/circuits/models/circuits.py:488` — **Medium** — `delete()` picks the pointer field from the in-memory `term_side`, not from the row it just locked

`delete()` issues `SELECT ... FOR UPDATE` on its own row (line 485) and then throws the result away, deriving the field name from `self.term_side` and the circuit from `self.circuit_id`. If either has diverged from the persisted row, `only_if_references=self.pk` correctly declines to clear the *wrong* pointer — but the pointer that genuinely references this termination is then never cleared by `_set_circuit_terminations`, so `on_delete=SET_NULL` nulls it with a bulk UPDATE and nothing is change-logged. That is exactly the gap #23134 is about, reintroduced through a side door.

Confirmed by repro, and codified by the PR's own `test_deletion_leaves_pointer_for_another_termination` (test_models.py:553): with `t_z.term_side` set to `'A'` in memory before `delete()`, `Circuit.termination_z` goes from the termination's PK to `NULL` and zero `ObjectChange` rows are written for the circuit. Since the row is already fetched under lock, using its persisted `circuit_id`/`term_side` — or simply clearing whichever of `termination_a`/`termination_z` equals `self.pk` — closes this at no extra cost.

### 2. `netbox/circuits/models/circuits.py:412` — **Medium** — the `save()` clear path has no `only_if_references` guard, so a stale instance clobbers another termination's pointer *and* change-logs the corruption

`delete()` gained `only_if_references=self.pk` so it only clears a pointer that actually references this object. The save path clears `termination_{_orig_term_side}` on `_orig_circuit_id` unconditionally, using values captured in `__init__` — i.e. at load time.

Confirmed repro: load `t1` (at C1/A) into `stale`; move `t1` to C2 via a second instance; create `t2` at (C1, A); then `stale.circuit = C3; stale.save()`. Result: `C1.termination_a` is nulled although it referenced the live `t2`, and `C2.termination_a` is left dangling at `t1`, which now lives on C3. The unconditional clear is pre-existing behaviour, but it is now wrapped in `snapshot()` + `save()`, so the corruption is written into the changelog as a legitimate user change and will be replayed by the branching plugin. Given the PR's own locking rationale ("the snapshot reflects a sibling pointer written concurrently"), the guard belongs on both paths.

### 3. `netbox/circuits/tests/test_models.py:539` — **Medium** — the cascade path is still unlogged, leaving the reported bug reachable by an ordinary UI action

`test_cascade_deletion_leaves_pointer_unrecorded` documents that deleting the terminating object (Site/Location/Region/ProviderNetwork) reaches the termination through the collector, so `delete()` is never called, `SET_NULL` clears `Circuit.termination_a`, and nothing is recorded. This is not an exotic path — deleting a site with circuits attached is routine, and a branching-plugin consumer replaying those records will drop the association exactly as described in #23134. The PR comment explains why a `pre_delete` receiver was rejected (`handle_deleted_object` connects earlier and would record the DELETE first), but that is solvable — e.g. handling it in `CustomCollector.delete()` (`netbox/netbox/models/deletion.py:130`), which already owns cascade ordering. As it stands, the fix covers two paths with tests and leaves a third hole with a test asserting the hole stays open.

### 4. `netbox/circuits/models/circuits.py:476` — **Medium** — bumping `Circuit.last_updated` invalidates the REST API's `If-Match` ETag

`NetBoxModelViewSet._get_etag()` (`netbox/netbox/api/viewsets/__init__.py:47`) derives the weak ETag purely from `last_updated`. Because `_set_circuit_terminations` does a full `circuit.save()`, creating/moving/deleting a *termination* now advances the parent `Circuit.last_updated`. Concrete failure: a client `GET`s a circuit, holds its ETag, someone adds a Z-side termination, and the client's `PATCH`/`DELETE` with `If-Match` fails with 412 Precondition Failed even though nothing it cares about changed. The circuit *list* ETag (`viewsets/__init__.py:281`) is invalidated the same way. Not necessarily wrong, but it's a user-visible API behaviour change that deserves a release note and a mention in the PR body alongside the other costs.

### 5. `netbox/circuits/models/circuits.py:476` — **Medium** — every termination write now fires `OBJECT_UPDATED` on the parent Circuit

`circuit.save()` runs `handle_changed_object` (`netbox/core/signals.py:88`), which both writes the `ObjectChange` (intended) and calls `enqueue_event(queue, circuit, request, OBJECT_UPDATED)` (side effect). Any event rule or webhook subscribed to Circuit updates now fires on every termination create, move, and delete; a bulk import of N terminations produces N circuit-update webhooks. The PR body notes this in passing — it needs to be in the release notes, since operators with circuit-update automation will see it as a regression.

### 6. `netbox/circuits/models/circuits.py:476` — **Low** — the full `circuit.save()` can silently backfill custom-field defaults onto the Circuit and attribute them to the termination write

The comment correctly explains why `update_fields` was avoided, but the consequence is that `CustomFieldsMixin.save()` (`netbox/netbox/models/features.py:343`) runs against the parent Circuit on every pointer write, populating defaults for any custom field not already present in `custom_field_data`. On an instance where a new custom field was added after the circuit was created, adding a termination writes those defaults into the circuit and folds them into the same `ObjectChange` as the pointer change — a data mutation the user did not ask for, recorded as if they had. Restricting the write to `update_fields=['termination_a', 'termination_z', 'custom_field_data', 'last_updated', ...]` would be narrower, or the snapshot could exclude the incidental fields.

### 7. `netbox/circuits/models/circuits.py:417,481` — **Low** — `using` is resolved without the `instance` router hint, unlike every other call site

Both new call sites use `router.db_for_write(type(self))`, while `DeleteMixin.delete()` (`netbox/netbox/models/deletion.py:139`) and Django's `Model.save()` both use `router.db_for_write(cls, instance=self)`. With a `DATABASE_ROUTERS` entry that consults the `instance` hint, `delete()` would open `transaction.atomic(using=X)` and lock/clear the pointer on database X while `super().delete()` deletes the row on database Y — the pointer write and the delete land in different transactions on different connections, defeating the atomicity the block exists to provide. Not reachable in a stock single-database NetBox, but trivial to make consistent.

**Comment by arthanson** [comment]
@claude /review Only report on issues of potential concern; we don't need commentary on anything that doesn't require further consideration.

**Comment by arthanson** [comment]
Per Claude review:

1. fixed
2. was resolved in changes to 1, no longer applies.
3. not fixed - this is a very, very narrow, silent edge case (custom field with a default, added after the circuit existed, backfill incomplete). Adds code complexity.
4. By design, we are doing a save which will have side effects
5. test suites cleaned up and narrowed previous overly broad tests.

File: netbox/netbox/models/deletion.py [hunk 1]
```diff
@@ -126,6 +126,18 @@ def collect(
                         # Add the model that the generic relation points to as a dependency
                         self.add_dependency(field.related_model, instance, reverse_dependency=True)
 
+    def delete(self):
+        # Clear any cached references to the objects being deleted first, so that each clear is
+        # recorded and precedes the DELETE. Django nulls a SET_NULL column with a bulk UPDATE,
+        # which emits no post_save and so is never change-logged. Models opt in by defining
+        # clear_cached_references(); cascaded objects reach this the same as explicit deletions.
+        with transaction.atomic(using=self.using):
```
Comment (jeremystretch) [inline netbox/netbox/models/deletion.py]: We should pass `savepoint=False` as Django's upstream `delete()` does.
```suggestion
        with transaction.atomic(using=self.using, savepoint=False):
```

---

Failure scenario: a UI bulk-delete of 500 devices previously issued zero savepoints inside the view's transaction and now issues 500. Past 64 subtransactions per backend, PostgreSQL's snapshot subxid cache overflows (`suboverflowed`), forcing pg_subtrans SLRU lookups and causing cluster-wide latency spikes for the duration. On the REST bulk path it doubles an already-per-object savepoint.

Passing `savepoint=False` gives the same "clears roll back with the DELETE" guarantee (it still opens a real transaction under autocommit) without the subtransaction.

File: netbox/netbox/models/deletion.py [hunk 2]
```diff
@@ -126,6 +126,18 @@ def collect(
                         # Add the model that the generic relation points to as a dependency
                         self.add_dependency(field.related_model, instance, reverse_dependency=True)
 
+    def delete(self):
+        # Clear any cached references to the objects being deleted first, so that each clear is
+        # recorded and precedes the DELETE. Django nulls a SET_NULL column with a bulk UPDATE,
+        # which emits no post_save and so is never change-logged. Models opt in by defining
+        # clear_cached_references(); cascaded objects reach this the same as explicit deletions.
```
Comment (jeremystretch) [inline netbox/netbox/models/deletion.py]: > cascaded objects reach this the same as explicit deletions

I'm not sure this is accurate: `CustomCollector` is called only by `DeleteMixin` below, whose `delete()` method is per-instance. AFAICT this won't catch deletions executed via a queryset (e.g. `circuit.terminations.all().delete()`). Not sure how large a concern that is, but probably worth evaluating further.

File: netbox/circuits/models/circuits.py [hunk 3]
```diff
@@ -241,6 +241,16 @@ def get_absolute_url(self):
...
 
 
+def _set_circuit_terminations(circuit, fields):
+    """
+    Set or clear a Circuit's cached `termination_a`/`termination_z` fields, recording the change.
+    """
+    circuit.snapshot()
+    for field_name, value in fields.items():
+        setattr(circuit, field_name, value)
+    circuit.save(update_fields=[*fields, 'last_updated'])
```
Comment (jeremystretch) [inline netbox/circuits/models/circuits.py:251]: Claude suggests we should include `custom_field_data` in the fields list. Seems like a niche risk but might not hurt.

---

`_set_circuit_terminations()` saves with `update_fields=[*fields, 'last_updated']`. `CustomFieldsMixin.save()` (`netbox/models/features.py:343-354`) mutates `self.custom_field_data` in place to inject defaults for custom fields still being provisioned, but `custom_field_data` is not in `update_fields`, so the injected values are never written — while `to_objectchange()` serializes the in-memory instance and *does* record them in `postchange_data`.

Failure scenario: a new `CustomField` on Circuit is mid-backfill when a termination is created. The resulting `ObjectChange.postchange_data` shows the circuit carrying the new field's default, but the DB row does not. A replay consumer (the netbox-branching case this PR exists for) would then apply a value that never existed.

Restricting the snapshot to the same fields, or including `custom_field_data` in `update_fields`, would keep the record honest.

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/netbox/models/deletion.py (hunk 2 above)
Comment (arthanson) [reply netbox/netbox/models/deletion.py]: Changed the comment, the underlying condition is pre-existing and systemic, so didn't change.

File: netbox/circuits/models/circuits.py (hunk 3 above)
Comment (arthanson) [reply netbox/circuits/models/circuits.py:251]: added with tests.
